Generate serialization of NSURLCredential

achristensen07 opened the pull request:
#### e1cab360c28ad35642b685bd17a9d3db6a052121
<pre>
Generate serialization of NSURLCredential
<a href="https://bugs.webkit.org/show_bug.cgi?id=269651">https://bugs.webkit.org/show_bug.cgi?id=269651</a>

Reviewed by Pascoe.

There&apos;s still work to be done to expose exactly what this dictionary should look like
in SerializedTypeInfo.mm somehow, but at least we&apos;re not using SecureCoding any more.
This is a baby step in the right direction.

This works on all platforms except macOS Monterey, which doesn&apos;t use
_enableStrictSecureDecodingMode for all types anyways, so on Monterey only, just use
NSKeyedUnarchiver.  Monterey also doesn&apos;t have the ability to serialize NSURLCredentials
with identities in them.  The Monterey code will be removed soon.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::encodeObjectDirectly&lt;NSObject&lt;NSSecureCoding&gt;&gt;):
(IPC::shouldEnableStrictMode):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.h: Added.
(WebKit::CoreIPCNSURLCredential::CoreIPCNSURLCredential):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.mm: Added.
(WebKit::CoreIPCNSURLCredential::CoreIPCNSURLCredential):
(WebKit::CoreIPCNSURLCredential::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(NSURLCredentialTesting_isEqual):
(operator==):
(TEST):

Canonical link: <a href="https://commits.webkit.org/275132@main">https://commits.webkit.org/275132@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9647c2eb0a2b661e51914f69e73c98d19bd2ac11

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40933 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19946 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43311 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43492 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37024 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43240 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22953 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17277 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33922 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41507 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16886 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35284 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14547 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14655 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36275 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44789 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/34386 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37140 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36585 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40330 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/40559 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15748 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12929 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17367 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/47570 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17418 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9765 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5454 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17011 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -530,6 +531,21 @@ static bool wkNSURLProtectionSpace_isEqual(NSURLProtectionSpace *a, SEL, NSURLPr
+        return false;
+    if (a.hasPassword != b.hasPassword)
+        return false;
+    if (a.certificates.count != b.certificates.count)
```
we should probably make sure the 
`@property (nullable, readonly) SecIdentityRef identity;`

matches.

pascoej replied:
also SecTrustRef. we have `secTrustRefsEqual` in this file to do it.

achristensen07 replied:
We don't have a way to compare SecIdentityRefs right now, but this functionality is tested in the test TestWebKitAPI.Challenge.ClientCertificate.

NSURLCredential has no accessor for a SecTrustRef, but NSURLCredential.credentialForTrust is also well tested.

achristensen07 replied:
I filed rdar://123342292

pascoej commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.serialization.in:
```diff
@@ -0,0 +1,34 @@
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#if HAVE(DICTIONARY_SERIALIZABLE_NSURLCREDENTIAL)
+[WebKitSecureCodingClass=NSURLCredential.class, SupportWKKeyedCoder] webkit_secure_coding NSURLCredential {
```
Can we omit `WebKitSecureCodingClass` since NSURLCredential doesn't require any softlinking to get at the class?

achristensen07 replied:
We can!

pascoej reviewed (approved):
Looks good, but had some questions / potential enhancements.

achristensen07 committed:
Generate serialization of NSURLCredential
https://bugs.webkit.org/show_bug.cgi?id=269651

Reviewed by Pascoe.

There's still work to be done to expose exactly what this dictionary should look like
in SerializedTypeInfo.mm somehow, but at least we're not using SecureCoding any more.
This is a baby step in the right direction.

This works on all platforms except macOS Monterey, which doesn't use
_enableStrictSecureDecodingMode for all types anyways, so on Monterey only, just use
NSKeyedUnarchiver.  Monterey also doesn't have the ability to serialize NSURLCredentials
with identities in them.  The Monterey code will be removed soon.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::encodeObjectDirectly<NSObject<NSSecureCoding>>):
(IPC::shouldEnableStrictMode):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.h: Added.
(WebKit::CoreIPCNSURLCredential::CoreIPCNSURLCredential):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.mm: Added.
(WebKit::CoreIPCNSURLCredential::CoreIPCNSURLCredential):
(WebKit::CoreIPCNSURLCredential::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(NSURLCredentialTesting_isEqual):
(operator==):
(TEST):

Canonical link: https://commits.webkit.org/275132@main
